flink/src/test/java/org/apache/iceberg/flink/data/TestHelpers.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    for (int i = 0; i < types.size(); i += 1) {
+      if (expected.get(i) == null) {
+        Assert.assertTrue(actual.isNullAt(i));
+        continue;
+      }
+      Type.TypeID typeId = types.get(i).typeId();
+      Object value = expected.get(i);
+      switch (typeId) {
+        case BOOLEAN:
+          Assert.assertEquals("boolean value should be equal", value, actual.getBoolean(i));
+          break;
+        case INTEGER:
+          Assert.assertEquals("int value should be equal", value, actual.getInt(i));
+          break;
+        case LONG:
+          Assert.assertEquals("long value should be equal", value, actual.getLong(i));
+          break;
+        case FLOAT:
+          Assert.assertEquals("float value should be equal", value, actual.getFloat(i));
+          break;
+        case DOUBLE:
+          Assert.assertEquals("double should be equal", value, actual.getDouble(i));
+          break;
+        case STRING:
+          Assert.assertTrue("Should expect a CharSequence", value instanceof CharSequence);
+          Assert.assertEquals("string should be equal", String.valueOf(value), actual.getString(i).toString());
+          break;
+        case DATE:
+          Assert.assertTrue("Should expect a Date", value instanceof LocalDate);
+          LocalDate date = ChronoUnit.DAYS.addTo(EPOCH_DAY, actual.getInt(i));
+          Assert.assertEquals("date should be equal", value, date);
+          break;
+        case TIME:
+          Assert.assertTrue("Should expect a LocalTime", value instanceof LocalTime);
+          int milliseconds = (int) (((LocalTime) value).toNanoOfDay() / 1000_000);
+          Assert.assertEquals("time millis should be equal", milliseconds, actual.getInt(i));
+          break;
+        case TIMESTAMP:
+          if (((Types.TimestampType) type.asPrimitiveType()).shouldAdjustToUTC()) {
+            Assert.assertTrue("Should expect a OffsetDataTime", value instanceof OffsetDateTime);
+            OffsetDateTime ts = (OffsetDateTime) value;
+            Assert.assertEquals("OffsetDataTime should be equal", ts.toLocalDateTime(),
+                actual.getTimestamp(i, 6).toLocalDateTime());
+          } else {
+            Assert.assertTrue("Should expect a LocalDataTime", value instanceof LocalDateTime);
+            LocalDateTime ts = (LocalDateTime) value;
+            Assert.assertEquals("LocalDataTime should be equal", ts,
+                actual.getTimestamp(i, 6).toLocalDateTime());
+          }
+          break;
+        case FIXED:
+          Assert.assertTrue("Should expect byte[]", value instanceof byte[]);
+          Assert.assertArrayEquals("binary should be equal", (byte[]) value, actual.getBinary(i));
+          break;
+        case BINARY:
+          Assert.assertTrue("Should expect a ByteBuffer", value instanceof ByteBuffer);
+          Assert.assertArrayEquals("binary should be equal", ((ByteBuffer) value).array(), actual.getBinary(i));
+          break;
+        case DECIMAL:
```

Comment:
Seems we share the same primitive value assertion between `assertRowData` and `assertArrayValues`, could it be a common method ?

Reply:
I tried to extract to one common method with an interface like `assertData(Type type, Record expected, Object actual` so that we can determine them via `instanceof`, but in the function, you have to call the getter with ((RowData)actual).getXXX and ((ArrayData)actual).getXXX.

Reply:
I tried to reuse those big switch-case,  How about the following way ? ( I did not finish all the code, but seems it should work to reuse that code). 

```java
public class TestHelpers {
  private TestHelpers() {
  }

  private static final OffsetDateTime EPOCH = Instant.ofEpochMilli(0L).atOffset(ZoneOffset.UTC);
  private static final LocalDate EPOCH_DAY = EPOCH.toLocalDate();

  public static void assertRowData(Type type, RowType rowType, Record expectedRecord, RowData actualRowData) {
    if (expectedRecord == null && actualRowData == null) {
      return;
    }

    Assert.assertTrue("expected Record and actual RowData should be both null or not null",
        expectedRecord != null && actualRowData != null);

    List<Type> types = new ArrayList<>();
    for (Types.NestedField field : type.asStructType().fields()) {
      types.add(field.type());
    }

    for (int i = 0; i < types.size(); i += 1) {
      if (expectedRecord.get(i) == null) {
        Assert.assertTrue(actualRowData.isNullAt(i));
        continue;
      }
      Object expected = expectedRecord.get(i);
      LogicalType logicalType = rowType.getTypeAt(i);

      final int fieldPos = i;
      assertEquals(types.get(i), rowType.getTypeAt(i), expected,
          () -> RowData.createFieldGetter(logicalType, fieldPos).getFieldOrNull(actualRowData));
    }
  }

  private static void assertEquals(Type type, LogicalType logicalType, Object expected, Supplier<Object> supplier) {
    switch (type.typeId()) {
      case BOOLEAN:
        Assert.assertEquals("boolean value should be equal", expected, supplier.get());
        break;
      case INTEGER:
        Assert.assertEquals("int value should be equal", expected, supplier.get());
        break;
      case LONG:
        Assert.assertEquals("long value should be equal", expected, supplier.get());
        break;
      case FLOAT:
        Assert.assertEquals("float value should be equal", expected, supplier.get());
        break;
      case DOUBLE:
        Assert.assertEquals("double value should be equal", expected, supplier.get());
        break;
      case STRING:
        Assert.assertTrue("Should expect a CharSequence", expected instanceof CharSequence);
        Assert.assertEquals("string should be equal",
            String.valueOf(expected), supplier.get().toString());
        break;
      case DATE:
        Assert.assertTrue("Should expect a Date", expected instanceof LocalDate);
        LocalDate date = ChronoUnit.DAYS.addTo(EPOCH_DAY, (int) supplier.get());
        Assert.assertEquals("date should be equal", expected, date);
        break;
      case TIME:
        Assert.assertTrue("Should expect a LocalTime", expected instanceof LocalTime);
        int milliseconds = (int) (((LocalTime) expected).toNanoOfDay() / 1000_000);
        Assert.assertEquals("time millis should be equal", milliseconds, supplier.get());
        break;
      case TIMESTAMP:
        if (((Types.TimestampType) type.asPrimitiveType()).shouldAdjustToUTC()) {
          Assert.assertTrue("Should expect a OffsetDataTime", expected instanceof OffsetDateTime);
          OffsetDateTime ts = (OffsetDateTime) expected;
          Assert.assertEquals("OffsetDataTime should be equal", ts.toLocalDateTime(),
              ((TimestampData) supplier.get()).toLocalDateTime());
        } else {
          Assert.assertTrue("Should expect a LocalDataTime", expected instanceof LocalDateTime);
          LocalDateTime ts = (LocalDateTime) expected;
          Assert.assertEquals("LocalDataTime should be equal", ts,
              ((TimestampData) supplier.get()).toLocalDateTime());
        }
        break;
      case BINARY:
        Assert.assertTrue("Should expect a ByteBuffer", expected instanceof ByteBuffer);
        Assert.assertArrayEquals("binary should be equal", ((ByteBuffer) expected).array(), (byte[]) supplier.get());
        break;
      case DECIMAL:
        Assert.assertTrue("Should expect a BigDecimal", expected instanceof BigDecimal);
        BigDecimal bd = (BigDecimal) expected;
        Assert.assertEquals("decimal value should be equal", bd,
            ((DecimalData) supplier.get()).toBigDecimal());
        break;
      case LIST: // TODO call assertArrayValues.
      case MAP:
      case STRUCT:
      case UUID:
      case FIXED:
        // TODO;
      default:
        throw new IllegalArgumentException("Not a supported type: " + type);
    }
  }

  private static void assertArrayValues(Type type, LogicalType logicalType, Collection<?> expectedArray,
                                        ArrayData actualArray) {
    List<?> expectedElements = Lists.newArrayList(expectedArray);
    for (int i = 0; i < expectedArray.size(); i += 1) {
      if (expectedElements.get(i) == null) {
        Assert.assertTrue(actualArray.isNullAt(i));
        continue;
      }

      Object expected = expectedElements.get(i);

      final int pos = i;
      assertEquals(type, logicalType, expected,
          () -> ArrayData.createElementGetter(logicalType).getElementOrNull(actualArray, pos));
    }
  }
}
```